Implement backend for FileJump (#8114)

masrlinu opened the pull request:
#### What is the purpose of this change?

This PR implements a backend for the storage provider FileJump.com.

#### Was the change discussed in an issue or in the forum before?

Yes, this backend was requested in issue #8114.

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [X] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

masrlinu commented:
Yes, everything works. I just tested it again. But I had to start it with a bigger timeout. I tested it with
test_all -backends filejump -timeout 10h

It finished in 1h48min.

Here is the output of the test:
[2025-07-21-140856.zip](https://github.com/user-attachments/files/21352390/2025-07-21-140856.zip)

ncw commented on backend/filejump/filejump_test.go:
```diff
@@ -0,0 +1,26 @@
+	fstests.Run(t, &fstests.Opt{
+		RemoteName:               "filejump:",
+		NilObject:                (*filejump.Object)(nil),
+		UnimplementableFsMethods: []string{"OpenWriterAt", "OpenChunkWriter"},
```
You can remove this line - it is only needed for Wrapping fs (eg crypt)

ncw commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1680 @@
+}
+
+
+// Check the interfaces are satisfied
```
Please put this at the end as that is where we look for it!

ncw commented on docs/content/filejump.md:
```diff
@@ -0,0 +1,215 @@
+
+### Modification times
+
+FileJump doesn't support setting custom modification times via the API. The modification time will be set to the upload time by the server. This means that `--update` and similar flags that rely on modification times may not work as expected.
```
Actually the `--update` flag is precisely for this situation.

Remove

> This means that `--update` and similar flags that rely on modification times may not work as expected.

and add

> You may want to use the [--update](/docs/#u-update) flag for a more accurate sync.

ncw commented on docs/content/filejump.md:
```diff
@@ -0,0 +1,215 @@
+
+### Restricted filenames
+
+FileJump has a minimum filename length requirement of 3 characters. Rclone will automatically pad shorter filenames with spaces to meet this requirement.
```
That is the first time I've seen that restriction in a cloud storage system!

I think this needs to be a little bit clearer

> Rclone will automatically pad shorter filenames with spaces on upload to meet this requirement. It will also remove the spaces on download. This means if you happen to have a 1 or 2 character file name ending in a space then syncing may go wrong.

ncw commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1680 @@
+	// FileJump doesn't support setting custom modification times via API
+	// The modification time is set by the server when the file is uploaded
+	// Return this error to indicate that setting mod time requires re-upload
+	return fs.ErrorCantSetModTimeWithoutDelete
```
You should just return `fs.ErrorCantSetModTime` here. The `fs.ErrorCantSetModTimeWithoutDelete` is for backends which can only set modification times when they upload a file.

ncw commented on docs/content/filejump.md:
```diff
@@ -0,0 +1,215 @@
+
+- FileJump doesn't support setting custom modification times
+- Upload time is used as the modification time
+- `--update` flag behavior may be affected
```
Using the `--update` flag can improve sync accuracy.

ncw commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1680 @@
+	iDir, _ := strconv.Atoi(rootID)
+	request := RequestDelete{
+		EntryIds:      []int{iDir},
+		DeleteForever: true,
```
Generally backends default to storing deletes in the trash and use a `hard_delete` flag to override.

Eg

      --jottacloud-hard-delete   Delete files permanently rather than putting them into the trash
      --mega-hard-delete         Delete files permanently rather than putting them into the trash

We started doing it that way round after one user deleted all of their files by accident! So safety first, but overridable by the user.

ncw commented on fstest/test_all/config.yaml:
```diff
@@ -162,6 +162,8 @@ backends:
    remote:   "TestGoFile:"
    fastlist: true
+ - backend:  "filejump"
+   remote:   "TestFileJump:"
```
Might want extra time here as discussed above!

ncw reviewed (changes requested):
This is looking really great. The integration tests all passing is fantastic.

I've put some notes inline - nothing serious.

I think you've got some lint things to fix too.

Thank you

masrlinu commented:
Hi @ncw! I've addressed all the suggestions, fixed the linter issues, updated the docs, and fixed the filejump-specific workspace handling. The latest changes also include support for move, dir-move, cleanup, about, and publiclink. Let me know what you think!

masrlinu commented on backend/filejump/api/types.go:
```diff
@@ -0,0 +1,119 @@
+func (i *Item) ModTime() (t time.Time) {
+	// Parse UpdatedAt first
+	if i.UpdatedAt != "" {
+		// Try multiple time formats that FileJump might use
```
Now we just use one format

masrlinu commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1680 @@
+	hasMetaData     bool
+	size            int64
+	modTime         time.Time
+	preserveModTime bool // Flag to indicate if modTime should be preserved
```
I removed it :-)

masrlinu commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1680 @@
+		Path:   path,
+		ExtraHeaders: map[string]string{
+			"Accept":          "application/json",
+			"Accept-Encoding": "gzip, deflate, br, zstd",
```
I changed it to just use gzip

ncw commented on docs/content/filejump.md:
```diff
@@ -0,0 +1,258 @@
+Storage> filejump
+** See help for filejump backend at: https://rclone.org/filejump/ **
+
+You should create an API access token here: https://drive.filejump.com/account-settings
```
This URL doesn't seem to work. It gives me "These credentials do not match our records"

https://eu.filejump.com/account-settings

Did work though.

However the token did not work giving these errors

```
2025/08/15 18:09:21 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2025/08/15 18:09:21 DEBUG : HTTP REQUEST (req 0xc00044a140)
2025/08/15 18:09:21 DEBUG : GET /api/v1/drive/file-entries?perPage=1000 HTTP/1.1
Host: drive.filejump.com
User-Agent: rclone/v1.71.0-beta.8915.df83968c9.pr-8693-filejump
Content-Length: 4
Accept: application/json
Accept-Encoding: gzip
Authorization: XXXX
Content-Type: application/json

null
2025/08/15 18:09:21 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2025/08/15 18:09:22 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2025/08/15 18:09:22 DEBUG : HTTP RESPONSE (req 0xc00044a140)
2025/08/15 18:09:22 DEBUG : HTTP/1.1 401 Unauthorized
Transfer-Encoding: chunked
Access-Control-Allow-Origin: *
Cache-Control: no-cache, private
Connection: keep-alive
Content-Type: application/json
Date: Fri, 15 Aug 2025 17:09:22 GMT
Server: nginx
Vary: Authorization

1e
{"message":"Unauthenticated."}
0
```

masrlinu replied:
Oh, I found out the reason. It is, because filejump has three different servers for different accounts. My account was an old one, so I also have to ask for the Server in the options.

ncw commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1969 @@
+		bodyJSON, _ := json.Marshal(body)
+		logMsg += fmt.Sprintf(" body=%s", string(bodyJSON))
+	}
+	fs.Debugf(f, "%v", logMsg)
```
Can you remove the debugging out of this function please.

You can see the transactions with `-vv --dump headers/bodies/requests/responses`

ncw commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1969 @@
+
+	if resp != nil && resp.StatusCode == 401 && strings.Contains(resp.Header.Get("Www-Authenticate"), "expired_token") {
+		authRetry = true
+		fs.Debugf(nil, "Should retry: %v", err)
```
No need for this debug - the pacer will debug

ncw commented on backend/filejump/filejump.go:
```diff
@@ -0,0 +1,1969 @@
+
+func (o *Object) upload(ctx context.Context, in io.Reader, leaf, directoryID string, size int64, modTime time.Time, options ...fs.OpenOption) (err error) {
+	// Upload via S3 presigned URL
+	fs.Debugf(o, "Uploading file via S3 presigned URL")
```
I'd probably remove this one too

ncw reviewed (commented):
I thought I'd give this a go so I went through the docs configuring it like a normal user.

I found a few things - see inline.

With the change for eu server I still couldn't get it to work though. It returned this on uploads

```
(error HTTP error 500 (500 Internal Server Error) returned body: "{\n    \"message\": \"There was an issue. Please try again later.\"\n}")
```

I was doing this test

```
rclone test makefiles --files 100 --max-file-size 2M /tmp/100files
rclone copy -Pvv /tmp/100files TestFileJump:100files
```

masrlinu commented:
So much work!! The server eu.filejump.com uses a different protocol for file storage. After a lot of effort, I finally got it working on both drive.filejump.com and eu.filejump.com. However, on eu.filejump.com two tests still fail:
```
            --- FAIL: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenSeek (0.42s)
            --- FAIL: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange (1.02s)
```
It seems that range reads/seek might not be supported there. For now, I’d like to mark this as unsupported in the current version and maybe add proper support in a future release. Is there a clean way to declare this as unsupported and have those tests skipped?

masrlinu commented:
I updated the backend as you suggested and also updated the documentation. All tests run via test_all passed for drive.filejump.com and eu.filejump.com. I couldn’t test it on the server app.filejump.com, since I don’t have an account there and can’t create one — account registration is only possible on the newest server, eu.filejump.com.

DWomack commented:
Just want to say thanks for your work.
I have it installed on two Pi and one amd64 ubuntu. 
I have 360 gig to move from drive.filejump.com to the new filejump.com via webdav and to filerule.com which is also webdav
I learned to use docker compose containers to build rclone.  Great lesson.
Thank you for the time you spent with this.

masrlinu commented:
You're welcome 😄 If you copy the data to two different clouds, instead of reading twice from the source, you can create a [union](https://rclone.org/union/), then you will have a backend, which acts like one but is actually two backends. You should configure it correctly, then if you copy files to there, they actually will be saved on both backends and are just read one time. And also for later, you can mount this union on your computer and the files will always be on both backends.
